sympy/tensor/tests/test_indexed.py

```diff
@@ -203,6 +203,26 @@ def test_IndexedBase_shape():
     raises(IndexException, lambda: F[i, j])
 
 
+def test_IndexedBase_assumptions():
+    i = Symbol('i', integer=True)
+    a = Symbol('a')
+    A = IndexedBase(a, positive=True)
```

Comment:
I think we should not be encouraging use of Indexed. The docstring for it should be changed from 
```
It is recommended that ``Indexed`` objects be created via ``IndexedBase``
```
to
```
It is recommended that ``Indexed`` objects be created by indexing ``IndexedBase``:
IndexedBase('A')[i, j] instead of Indexed(IndexedBase('A'), i, j).
```
This class is created when IndexedBase has a `getitem` applied in the same way that `Function('f')` which is an Undef becomes AppliedUndef when it is called as `Function('f')(x)`. Functions are declared real at the Function level (`f = Function('f', real=True)` and then `f(x).is_real` -> True irrespective of what assumptions that `x` has in the same way that `f.is_real` -> True). 

In similar manner, I would expect that for `A = IndexedBase('A', real=True)` then we would have `A.is_real` and `A['i'].is_real`->True. A natural way to avoid having to split keywords is to require that IndexedBase inherits assumptions from the Symbol, if provided. We could do the same for Function, i.e. Function(Symbol('f', real=True)).is_real -> True. In such cases, however, I would raise an error if assumptions were passed when a Symbol was used: if the user doesn't want the Symbol's assumptions they could pass Symbol.name.

The main-point summary is that I don't think we should have `Indexed(IndexedBase('A'), i, j, real=True)`. The assumptions should be associated with IndexedBase, optionally inherited from the Symbol used to provide the name (but if that is done it should probably be done with Function, too).

Reply:
Instantiation of `IndexBase` could go as
```python
base = sympify(base)
assumptions = base.assumptions0
passed_assumptions, kwargs = split assumptions
sanitize passed_assumptions
for k, v in passed_assumptions.items():
    if v is None:  # can it be?
        continue
    vis = assumption.get(k, None)
    if vis is None:
        assumptions[k] = v
    elif v != vis:
        raise ValueError('clash between base and passed assumptions')
    else:
        pass  # v == vis
# assumptions are ready to store
```

Reply:
To be clear, you want to remove support for assumptions to be specified directly in the constructor of `Indexed` (with appropriate changes to the docstring), so that the only way to add assumptions is to set it on the `IndexedBase` object or inherit from the symbol used to instantiate `IndexedBase`? I'll be happy to do that, but just want to make sure about the desired changes. 

Should the part about inheriting from symbol be included in this PR, or should another one be made for adding this to both `Function` and `IndexedBase`? Could change `Function` in this PR also, but that seems less clear.

PS: Sorry about the delayed response.

Reply:
Yes, "only way to add assumptions is to set it on the IndexedBase".

And it makes sense to make inheritance from a Symbol a separate PR, handling Function at the same time.

Reply:
Done. I'll open a new PR for inheritance from Symbol if/when this is accepted.